src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs

```diff
@@ -700,32 +700,71 @@ public void ExecuteCodeBlockActions<TLanguageKindEnum>(
-                codeBlockStartActions, codeBlockActions, codeBlockEndActions, analyzer,
-                declaredNode, declaredSymbol, executableCodeBlocks, (codeBlocks) => codeBlocks.SelectMany(
-                    cb =>
+            Debug.Assert(!executableCodeBlocks.IsEmpty);
+            var executableNodeActions = ArrayBuilder<SyntaxNodeAnalyzerAction<TLanguageKindEnum>>.GetInstance();
+            ExecuteBlockActionsCore(
+                codeBlockStartActions,
+                codeBlockActions,
+                codeBlockEndActions,
+                analyzer,
+                declaredNode,
+                declaredSymbol,
+                operationBlocks: ImmutableArray<IOperation>.Empty,
+                semanticModel,
+                filterSpan,
+                isGeneratedCode,
+                executableNodeActions,
+                addActions: static (startAction, codeBlockEndActions, syntaxNodeActions, args, cancellationToken) =>
+                {
+                    var (@this, codeBlockStartActions, analyzer, declaredNode, declaredSymbol, executableCodeBlocks, semanticModel, getKind, filterSpan, isGeneratedCode) = args;
+
+                    var codeBlockScope = new HostCodeBlockStartAnalysisScope<TLanguageKindEnum>(startAction.Analyzer);
+                    var blockStartContext = new AnalyzerCodeBlockStartAnalysisContext<TLanguageKindEnum>(
+                        codeBlockScope, declaredNode, declaredSymbol, semanticModel, @this.AnalyzerOptions, filterSpan, isGeneratedCode, cancellationToken);
+
+                    // Catch Exception from the start action.
+                    @this.ExecuteAndCatchIfThrows(
+                        startAction.Analyzer,
+                        static data =>
+                        {
+                            data.action(data.context);
+                            data.blockEndActions?.AddAll(data.scope.CodeBlockEndActions);
+                            data.syntaxNodeActions?.AddRange(data.scope.SyntaxNodeActions);
+                        },
+                        (action: startAction.Action, context: blockStartContext, scope: codeBlockScope, blockEndActions: codeBlockEndActions, syntaxNodeActions),
+                        new AnalysisContextInfo(@this.Compilation, declaredSymbol, declaredNode),
+                        cancellationToken);
+                },
+                executeActions: static (syntaxNodeActions, diagReporter, isSupportedDiagnostic, args, cancellationToken) =>
+                {
+                    var (@this, codeBlockStartActions, analyzer, declaredNode, declaredSymbol, executableCodeBlocks, semanticModel, getKind, filterSpan, isGeneratedCode) = args;
+
+                    Debug.Assert(getKind != null);
+
+                    var executableNodeActionsByKind = GetNodeActionsByKind(syntaxNodeActions);
```

Comment:
>GetNodeActionsByKind [](http://example.com/codeflow?start=54&length=20)

Trying to piece this together, but GetNodeActionsByKind doesn't need to take in an IEnumerable anymore right?

Reply:
it does.  its' called from two locations still (one in AnalyzeExecutor), one which is an IA, one which is an AB.